Prepare KubeVirt for multi-version api support.

davidvossel opened the pull request:
This PR prepares KubeVirt for multi-version api support. 

```release-note
NONE
```

davidvossel commented:
The PR has been rebased and the failing tests should be addressed at this point.

davidvossel commented:
/retest

rmohr commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -159,6 +159,15 @@ func (c *MigrationController) execute(key string) error {
 
+	// this must be first step in execution. Writting the object
+	// when api version changes ensures our api stored version is updated.
+	if !controller.ObservedLatestApiVersionAnnotation(migration) {
```
What happens if a user fetches a VMI which was not yet processed and tries to update it with the old version and we didn't to a write?

davidvossel replied:
it's fine. it doesn't actually matter who does the write or with what version. We just need to ensure a write occurs after an api change occurs. That's the intent behind this code block.

The user can't impact the storage api version. If they update the vmi using v1alpha3 or using v1, it doesn't matter internally.  it is functionally the exact same.

rmohr replied:
What happens if there was no write?

davidvossel replied:
It depends.

If the Storage version changed, then we aren't guaranteed the object is written internally to use the new Storage version.

If only a new API version is introduced, any attempt to PATCH the existing object using a client with the new API version will fail until the object is UPDATED using the previous client.

rmohr replied:
This is really so ugly did we make sure that there is a bug report regarding to this?

davidvossel commented:
> Can be outlined how the update flow exactly is and what the impact on availability is? It looks a little bit like we have to write with the new apiversion, this may have impact on availability.

There's very little impact on availability. The only impact is that the VM/VMI UPDATE must occur before a VM/VMI can be PATCHed using a newly introduced api versioned client. We're just writing an annotation to reflect the known latest and storage api versions to ensure a write occurs when either of those change.

so, it only impacts user's ability to modify objects using PATCH with the a client using the new api version. It would be surprising if anyone hit this because it would require the user to attempt to use a client with the newly introduced api version to issue a PATCH before the controller could write that annotation.

davidvossel commented:
> I have one request: There is this test: https://github.com/kubevirt/kubevirt/blob/master/tests/subresource_api_test.go#L223. It ensures that we are properly registered on the apiserver. Can we run it also after the update? It's pretty important that this registration works without issues, because many dynamic clients just fail if one endpoint returns an error.

done.

rmohr commented on tests/operator_test.go:
```diff
@@ -623,6 +623,13 @@ spec:
+				spec, err := virtClient.RestClient().Get().AbsPath("/openapi/v2").DoRaw()
+				Expect(err).ToNot(HaveOccurred())
+				return string(spec)
+			}, 60*time.Second, 1*time.Second).Should(ContainSubstring("subresources.kubevirt.io"))
```
Can we check for both versions?

rmohr replied:
:+1:

slintes commented on pkg/virt-operator/kubevirt.go:
```diff
@@ -546,6 +555,10 @@ func (c *KubeVirtController) generateInstallStrategyJob(kv *v1.KubeVirt) *batchv
 									Value: string(pullPolicy),
 								},
+								{
+									Name:  util.TargetImagePullPolicy,
```
duplicate from above?

slintes replied:
ping

davidvossel replied:
yep, this was addressed

slintes commented:
heads up, did not test it yet, but AFAIU we will run into this k8s issue [0]
It is fixed in 1.14 [1], but not in 1.13 yet [2]

[0] https://github.com/kubernetes/kubernetes/issues/73752
[1] https://github.com/kubernetes/kubernetes/pull/74154
[2] https://github.com/kubernetes/kubernetes/pull/79495

davidvossel commented:
> heads up, did not test it yet, but AFAIU we will run into this k8s issue [0]
It is fixed in 1.14 [1], but not in 1.13 yet [2]
> [0] kubernetes/kubernetes#73752
>[1] kubernetes/kubernetes#74154
>[2] kubernetes/kubernetes#79495 

We don't seem to be hitting [0] now. I think my fix related to writing the objects after a new api version is working around the issue... because without writing the object, i saw the exact same error [0] is reporting. But after introducing logic to perform a write on every object, that error went away.

davidvossel commented:
> StorageGroupVersion / SubresourceStorageGroupVersion: I see that these are used in the kubecli package, but do not understand their meaning and why they stay at v1alpha3?

mostly for n-1 KubeVirt release compatiblity. We're guaranteed the latest "storage" version is going to at least be compatible with the previous version of KubeVirt because we don't introduce a new api version and immediately make it the default storage version as well.

davidvossel commented:
@slintes @rmohr can i get another review round, or at least lgtm/approved labels.  I didn't make any new changes. Just rebased to latest master and fixed some merge conflicts.

rmohr commented:
/test all

rmohr commented:
/retest

rmohr commented:
/retest

rmohr commented:
/retest

rmohr commented:
@davidvossel one last thing as a follow-up for the off-issue mailing discussion regarding to existing bugs on CR conversion: There was the issue which gets only triggerd if the k8s apiserver (not kubevirt apiserver) is restarted. Are you sure we can't hit that?

davidvossel commented:
> There was the issue which gets only triggerd if the k8s apiserver (not kubevirt apiserver) is restarted. Are you sure we can't hit that

yeah, we definitely hit that like everyone else. I don't have any suggestions there other than k8s needs to fix it.

davidvossel commented:
> We don't seem to be hitting [0] now. I think my fix related to writing the objects after a new api version is working around the issue... because without writing the object, i saw the exact same error [0] is reporting. But after introducing logic to perform a write on every object, that error went away.

i was able to reproduce https://github.com/kubernetes/kubernetes/issues/73752 with older versions of k8s.

davidvossel commented:
@rmohr @slintes @fabiand 

As a result of https://github.com/kubernetes/kubernetes/issues/73752 this PR has been modified to only prepare KubeVirt for multi-version support. The goal here is to get the heavy lifting in this PR merged without actually introducing the new api version. Once we feel comfortable k8s has all the support we need for multiple versions, having this preparation work already out of the way should make it easy for us to add a new version. 

Essentially I don't feel comfortable merging a new version until the k8s fix I referenced is fixed in at least k8s 1.13.x. There is already a PR open (https://github.com/kubernetes/kubernetes/pull/79495) to backport the k8s fix to 1.13, so hopefully there will be a 1.13 release soon that works.

slintes commented:
/test all

slintes commented:
rebased for getting better test results...

slintes commented:
Only very few unrelated test failures.
@rmohr can you have another look please? I would be fine with merging.

/lgtm

rmohr commented:
looking. @slintes thanks for the preparations.

rmohr commented on Makefile:
```diff
@@ -35,8 +35,8 @@ generate:
 
-client-python:
-	hack/dockerized "./hack/generate.sh && TRAVIS_TAG=${TRAVIS_TAG} ./hack/gen-client-python/generate.sh"
+client-python: generate
```
This is not the right direction. Intentionally the logic should be in the bash scripts to avoid starting and stopping the containers multiple times because for many engineers it makes the difference between fast and seamless execution and having to wait for 5-10 seconds longer. Not so important for this target, but important in general. We can fix it in another PR.

davidvossel commented on pkg/virt-api/api.go:
```diff
@@ -186,151 +189,161 @@ func subresourceAPIGroup() metav1.APIGroup {
 		To(func(request *restful.Request, response *restful.Response) {
+			paths := []string{"/apis",
+				"/apis/",
+				rest.GroupBasePath(v1.SubresourceGroupVersions[0]),
```
this was fixed
